flang/lib/Lower/FIRBuilder.cpp

```diff
@@ -46,11 +46,13 @@ mlir::Type Fortran::lower::FirOpBuilder::getVarLenSeqTy(mlir::Type eleTy) {
 }
 
 mlir::Value
-Fortran::lower::FirOpBuilder::createNullConstant(mlir::Location loc) {
+Fortran::lower::FirOpBuilder::createNullConstant(mlir::Location loc,
+                                                 mlir::Type ptrType) {
   auto indexType = getIndexType();
   auto zero = createIntegerConstant(loc, indexType, 0);
-  auto noneRefType = getRefType(getNoneType());
-  return createConvert(loc, noneRefType, zero);
+  if (!ptrType)
+    ptrType = getRefType(getNoneType());
```

Comment:
Is this just to avoid a `fir.convert` that's going to be optimized away?

Reply:
They are optimized out at some point of the chain, but they are not yet in fir and I find that the chains of fir.convert makes the the fir output harder to proof read. I think the cost for a little bit more of fir readability is not too high here.

I do not like writing lit tests where you have to go through zero -> fir.ref -> fir.ptr just to check that the fir.ptr is set to zero. Some convert have semantic impact (truncation/extension), so it's not safe to assume you can just always ignore them when proof reading/testing fir outputs.